### PR TITLE
Added Horspool algorithm

### PR DESCRIPTION
**Pull request by ckontakis**
#### Checklist

- [x ] Added description of change
- [x ] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x ] Added tests and example, test must pass
- [x ] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x ] Relevant documentation/comments is changed or added
- [x ] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x ] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x ] I acknowledge that all my contributions will be made under the project's license.

Notes: 
An implementation of Horspool's algorithm for finding substrings in strings

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1071"></a>

### CODE REVIEW

File: strings/CMakeLists.txt [hunk 1]
```diff
@@ -7,7 +7,7 @@ file( GLOB APP_SOURCES RELATIVE ${CMAKE_CURRENT_SOURCE_DIR} *.cpp )
 foreach( testsourcefile ${APP_SOURCES} )
     # I used a simple string replace, to cut off .cpp.
     string( REPLACE ".cpp" "" testname ${testsourcefile} )
-    add_executable( ${testname} ${testsourcefile} )
+    add_executable( ${testname} ${testsourcefile} horspool.cpp)
```
Comment (kvedala) [inline strings/CMakeLists.txt:10]: No changes are required in the CMake file. Please undo this.

Comment (ckontakis) [reply strings/CMakeLists.txt]: Thanks for the advice!

File: strings/horspool.cpp [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+            }
+        }
+    }
+    return shiftTable;
+}
+
+/**
+ * A function that implements Horspool's algorithm
+ * @param text
+ * @param prototype
```
Comment (Panquesito7) [inline strings/horspool.cpp:38]: Same here as above.

File: strings/horspool.cpp [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    assert(horspool("Hello World","ello") == true);
+    assert(horspool("Hello World","rld") == true);
+    assert(horspool("Hello","Helo") == false);
+    assert(horspool("c++_algorithms","c++_algorithms") == true);
+    assert(horspool("c++_algorithms","c++_") == true);
+    assert(horspool("Hello","Hello World") == false);
+    assert(horspool("c++_algorithms","") == false);
+    assert(horspool("c++","c") == true);
+    assert(horspool("3458934793","4793") == true);
+    assert(horspool("3458934793","123") == false);
```
Comment (Panquesito7) [inline strings/horspool.cpp:87]: Please use test implementations in a separate function.
Confused? See #962 as a reference. 🙂

File: strings/horspool.cpp [hunk 4]
```diff
@@ -0,0 +1,89 @@
+/**
+ * @file
+ * @brief Horspool's algorithm that finds if a string contains a substring (https://en.wikipedia.org/wiki/Boyer%E2%80%93Moore%E2%80%93Horspool_algorithm)
+ * @author [Harry Kontakis] (https://github.com/ckontakis)
```
Comment (Panquesito7) [inline strings/horspool.cpp:4]:
```suggestion
 * @author [Harry Kontakis](https://github.com/ckontakis)
```

File: strings/horspool.cpp [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+ */
+
+#include <iostream>
+#include <unordered_map>
+#include <cassert>
+
+/**
```
Comment (Panquesito7) [inline strings/horspool.cpp:11]:
```suggestion
/**
 * @namespace strings
 * @brief Algorithms with strings
 */
namespace strings {
/**
 * @namespace horspool
 * @brief Functions for [Horspool's](https://en.wikipedia.org/wiki/Boyer%E2%80%93Moore%E2%80%93Horspool_algorithm) algorithm
 */
namespace horspool {
/**
```

File: strings/horspool.cpp [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+        }else{
+            if(shiftTable.find(text[i]) != shiftTable.end()){
+                i += shiftTable[text[i]]; // If shift table contains the character then shift index as many steps as value
+            }else{
+                i += prototype.size(); // If character does not exist in shift table then shift index as many steps as size of prototype string
+            }
+        }
+    }
+    return false;
+}
```
Comment (Panquesito7) [inline strings/horspool.cpp:70]:
```suggestion
}
}  // namespace horspool
}  // namespace strings

```

File: strings/horspool.cpp (hunk 3 above)
Comment (Panquesito7) [inline strings/horspool.cpp:87]:
```suggestion
    assert(strings::horspool::horspool("Hello World","World") == true);
    assert(strings::horspool::horspool("Hello World"," World") == true);
    assert(strings::horspool::horspool("Hello World","ello") == true);
    assert(strings::horspool::horspool("Hello World","rld") == true);
    assert(strings::horspool::horspool("Hello","Helo") == false);
    assert(strings::horspool::horspool("c++_algorithms","c++_algorithms") == true);
    assert(strings::horspool::horspool("c++_algorithms","c++_") == true);
    assert(strings::horspool::horspool("Hello","Hello World") == false);
    assert(strings::horspool::horspool("c++_algorithms","") == false);
    assert(strings::horspool::horspool("c++","c") == true);
    assert(strings::horspool::horspool("3458934793","4793") == true);
    assert(strings::horspool::horspool("3458934793","123") == false);
```

File: strings/horspool.cpp [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+} // namespace horspool
+} // namespace strings
+
+/**
+ * @brief Function with test cases for Horspool's algorithm
+ */
+void test(){
```
Comment (Panquesito7) [inline strings/horspool.cpp:99]:
```suggestion
static void test(){
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

File: strings/horspool.cpp [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+            }
+        }
+    }
+    return false;
+}
+} // namespace horspool
+} // namespace strings
+
+/**
+ * @brief Function with test cases for Horspool's algorithm
```
Comment (Panquesito7) [inline strings/horspool.cpp:97]:
```suggestion
 * @brief Function with test cases for Horspool's algorithm
 * @returns void
```

File: strings/horspool.cpp [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+ * @author [Harry Kontakis](https://github.com/ckontakis)
+ */
+
+#include <iostream>
+#include <unordered_map>
+#include <cassert>
+
+/**
+ *@brief Horspool's algorithm with strings.
+ */
```
Comment (Panquesito7) [inline strings/horspool.cpp:13]:
```suggestion
/**
 * @namespace strings
 * @brief Algorithms with strings
 */
```

File: strings/horspool.cpp [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+namespace strings {
+
+/**
+ *@brief An implementation of Horspool's algorithm with functions.
+ */
```
Comment (Panquesito7) [inline strings/horspool.cpp:18]:
```suggestion
/**
 * @namespace horspool
 * @brief Functions for [Horspool's](https://en.wikipedia.org/wiki/Boyer%E2%80%93Moore%E2%80%93Horspool_algorithm) algorithm
 */
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

File: strings/horspool.cpp [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+#include <iostream>
+#include <unordered_map>
+#include <cassert>
+
+/**
+ * @namespace strings
+ * @brief Algorithms with strings
+ */
+namespace strings {
+
```
Comment (Panquesito7) [inline strings/horspool.cpp:16]:
```suggestion
```

File: strings/horspool.cpp [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+/**
+ * @namespace horspool
+ * @brief Functions for [Horspool's](https://en.wikipedia.org/wiki/Boyer%E2%80%93Moore%E2%80%93Horspool_algorithm) algorithm
+ */
+namespace horspool {
+
```
Comment (Panquesito7) [inline strings/horspool.cpp:22]:
```suggestion
```

**Review by Panquesito7 — APPROVED** [review]
Good work; LGTM. 👍 😄
Thanks for your contribution.

@kvedala @ayaankhan98 please review as well. Thanks. 🙂

**Review by ayaankhan98 — APPROVED** [review]
good job, LGTM.

**Pull request merged**